udtf.go

```diff
@@ -0,0 +1,446 @@
+
+		var err error
+		namedArgs[name], err = getValue(t, value)
+		if err != nil {
+			setBindError(info, err.Error())
+			return
+		}
+	}
+
+	instance, err := tfunc.BindArguments(namedArgs, args...)
+	if err != nil {
+		setBindError(info, err.Error())
+		return
+	}
+
+	columns := instance.Columns()
+
+	instanceData := tableFunctionData{
+		fun:        instance,
+		projection: make([]int, len(columns)),
+	}
+
+	for i, v := range columns {
+		dt := v.T.logicalType()
+		defer C.duckdb_destroy_logical_type(&dt)
+		colName := C.CString(v.Name)
+		defer C.duckdb_free(unsafe.Pointer(colName))
+		C.duckdb_bind_add_result_column(info, colName, dt)
+
+		instanceData.projection[i] = -1
+	}
+
+	cardinality := instance.Cardinality()
+	if cardinality != nil {
+		C.duckdb_bind_set_cardinality(info, C.idx_t(cardinality.Cardinality), C.bool(cardinality.Exact))
+	}
+
+	pinnedInstanceData := pinnedValue[tableFunctionData]{
+		pinner: &runtime.Pinner{},
+		value:  instanceData,
+	}
+
+	handle := cgo.NewHandle(pinnedInstanceData)
+	pinnedInstanceData.pinner.Pin(&handle)
+	C.duckdb_bind_set_bind_data(info, unsafe.Pointer(&handle), C.duckdb_delete_callback_t(C.udf_delete_callback))
+}
+
+//export table_udf_init
+func table_udf_init(info C.duckdb_init_info) {
+	instance := getPinned[tableFunctionData](C.duckdb_init_get_bind_data(info))
+	instance.setColumnCount(info)
+	instance.fun.(tableSource).Init()
+}
+
+//export table_udf_init_threaded
+func table_udf_init_threaded(info C.duckdb_init_info) {
+	instance := getPinned[tableFunctionData](C.duckdb_init_get_bind_data(info))
+	instance.setColumnCount(info)
+	initData := instance.fun.(threadedTableSource).Init()
+	C.duckdb_init_set_max_threads(info, C.idx_t(initData.MaxThreads))
+}
+
+//export table_udf_local_init
+func table_udf_local_init(info C.duckdb_init_info) {
+	instance := getPinned[tableFunctionData](C.duckdb_init_get_bind_data(info))
+	localState := pinnedValue[any]{
+		pinner: &runtime.Pinner{},
+		value:  instance.fun.(threadedTableSource).NewLocalState(),
+	}
+	handle := cgo.NewHandle(localState)
+	localState.pinner.Pin(&handle)
+	C.duckdb_init_set_init_data(info, unsafe.Pointer(&handle), C.duckdb_delete_callback_t(C.udf_delete_callback))
+}
+
+//export table_udf_row_callback
+func table_udf_row_callback(info C.duckdb_function_info, output C.duckdb_data_chunk) {
+	instance := getPinned[tableFunctionData](C.duckdb_function_get_bind_data(info))
+
+	var chunk DataChunk
+	err := chunk.initFromDuckDataChunk(output, true)
+	if err != nil {
+		setFuncError(info, err.Error())
+		return
+	}
+
+	row := Row{
+		chunk:      &chunk,
+		projection: instance.projection,
+	}
+
+	maxSize := C.duckdb_vector_size()
+
+	switch fun := instance.fun.(type) {
+	case RowTableSource:
+		// At the end of the loop row.r must be the index one past the last added row
+		for row.r = 0; row.r < maxSize; row.r++ {
+			nextResults, err := fun.FillRow(row)
+			if err != nil {
+				setFuncError(info, err.Error())
+			}
```

Comment:
Should we abort if we encounter an error?
Same for the `ThreadedRowTableSource` case.